Add more personal information fields to the settings page for enhance…

ChristophWurst opened the pull request:
…d federated sharing

TODO:
- [x] fill in data on page load
- [x] fix popover position for different text lengths
- [x] fix popover position for the 'full name' field
- [x] show read-only fields when editing is not allowed (fix todo in template file)
- [x] ajax calls that save settings
- [x] persist federated sharing scopes (private, contacts, public)
- [ ] find appropriate place for the textual feedback after an successful/unsuccessful ajax call. cc @jancborchardt 
- [x] user loading/success feedback when changing scopes
- [x] add avatar scope
- [x] check whether removed routes were used elsewhere
  - [ ] `settings/js/user/user.js` uses the /mailAddress route
  - [ ] `settings/js/user/user.js` uses the /displayName route
- [ ] check/fix avatar updating once settings are persisted on the server, so the displayname can be changed again.
- [ ] add js unit tests
- [ ] integrate sync of the personal data into the system addressbook
- [ ] adjust "get email address" / "get display name" implementation to first look at the accounts table and fall back to the old location if needed

@schiesbn 

![bildschirmfoto von 2016-04-20 17-49-10](https://cloud.githubusercontent.com/assets/1374172/14681141/300b4444-0720-11e6-832a-3162c9351245.png)

ChristophWurst committed:
Add more personal information fields to the settings page for enhanced federated sharing

fix layout

Add generic way of handling input change events

DeepDiver1975 commented:
> ajax calls that save settings

@ChristophWurst please ping me on irc as soon as you start working on this one so that we can discuss on how the api will look like.

My idea: Have one route where we get and set all personal information at once and remove the calls like set displayname and set email ....

THX

ChristophWurst committed:
Persist settings on the server

Persist personal settings federated sharing scopes

Show new settings fields in read-only mode too

Insert values on page load

Return updated values; show inline success feedback

ChristophWurst committed:
Add federation scope to the user avatar

ChristophWurst commented:
@DeepDiver1975 @schiesbn I have to adjust the user list js to use the new route too, as it's using those two removed ones. As not all of the information is being POSTed there, so how about using PATCH instead and allow partial updates?

ChristophWurst committed:
use PUT to update user settings

schiesbn commented:
@DeepDiver1975 any thoughts about where to store the data? Should we add to the oc_users table a vcard column? Or a complete new table? But a additional table could become quite confusing.

schiesbn commented:
Or a complete different idea... Because we are using vcards anyway, the best solution is probably to create a new address book for the personal infos... @DeepDiver1975 what do you think?

jancborchardt commented:
Why a new address book? What others like Android and iOS do is create a contact for yourself in the normal address book. :)

schiesbn commented:
> Why a new address book? What others like Android and iOS do is create a contact for yourself in the normal address book. :)

I assume with the "normal address book" you mean the system wide address book we already have. We can't use this to collect the information because it will synced to trusted ownClouds and we don't want to sync data which was set to "private". Therefore we need two system address books, one which contain all the information from all users, and one which contain only the information the user want to share.

schiesbn commented:
While thinking about it, we could also store all data on the users "default" address book, in case something like this exists. And every time we update the users defaults address book we make sure that we keep the look-up server and the system wide address book in sync. This way we might be able to avoid a seconds system wide address book.

DeepDiver1975 commented:
Storing the personal data as vcard has first to be verified.

Where ever we store that information - the data will be synced to the system addressbook based on the visibility as defined. This mechanism is already in place.

Storing the personal information in the users default addressbook might make sense - there is even a way to identify the so called me-card http://sabre.io/dav/carddav-me-card/

DeepDiver1975 commented:
As for this pr I think it is enough to store the information within the oc_preferences the same way as we do with displayname and email.

DeepDiver1975 commented:
But let's no add one field for each property but save one json for now

schiesbn commented:
> As for this pr I think it is enough to store the information within the oc_preferences the same way as we do with displayname and email.
> But let's no add one field for each property but save one json for now

OK, let's do it this way. This also makes it a lot easier for the first iteration.

DeepDiver1975 commented:
*one should hit the 'comment' button*

Because of the additional information regarding the visibility of each field I don't know if vcard is the best format for storing this information - I need to think about this.

Regarding the location of storage - I'm still in favor of using the user_account_table as discussed.

Commit:
introduce accounts table and keep it up-to-date with the data added to the personal settings

schiesbn commented:
@DeepDiver1975 please have a look at my last commit, this introduced a 'accounts' table.

DeepDiver1975 commented on settings/Controller/UsersController.php:
```diff
@@ -509,7 +517,22 @@ public function saveUserSettings($userId, $avatarScope,
+		$data = [
+			'avatar' =>  ['scope' => $avatarScope],
+			'displayname' => ['value' => $displayname, 'scope' => $displaynameScope],
+			'email' => ['value' => $email, 'scope' => $emailScope],
```
let's design this in a way so that multiple can easily be stored in the future for email, website, address and phone.

DeepDiver1975 commented on db_structure.xml:
```diff
@@ -1723,4 +1723,38 @@
+			</field>
+			<field>
+				<name>data</name>
+				<type>text</type>
```
clob - not text(4000)

DeepDiver1975 commented:
> integrate sync of the personal data into the system addressbook

just adding a new todo

DeepDiver1975 commented on settings/Application.php:
```diff
@@ -104,6 +105,9 @@ public function __construct(array $urlParams=[]){
 			);
 		});
+		$container->registerService('AccountManager', function(IContainer $c) {
+			return new AccountManager(\OC::$server->getDatabaseConnection());
```
should be taken from $c

````php
$c->getContainer()-getServer()->getDatabaseConnection()

LukasReschke commented on settings/Controller/UsersController.php:
```diff
@@ -484,16 +492,78 @@ public function destroy($id) {
+					$address, $addressScope) {
+
+		if($userId === null) {
+			$userId = $this->userSession->getUser()->getUID();
```
Shouldn't there be like … a permission check somewhere here? 🙈

DeepDiver1975 replied:
and a null check is missing from my pov

schiesbn replied:
> Shouldn't there be like … a permission check somewhere here?

We will see if we need the `$userId` parameter at all. At the moment I think we can remove it... let's see what will happen if development moves forward

> nd a null check is missing from my pov

The method can only be called if a user is logged in (no `@PublicPage`) -> we always have a valid user

DeepDiver1975 replied:
> The method can only be called if a user is logged in (no @PublicPage) -> we always have a valid user

then we better pass in the user id - no need to pass in the session

schiesbn replied:
It was already there and used in many places, so I re-use it instead of adding another parameter to the constructor (which already has 16!). At this point I don't want to re-write the complete userController but just extend/modify it to do what I need.

DeepDiver1975 replied:
Oh yes - I see your point ...

Commit:
display stored user data

Commit:
allow multiple values for phone, website, address and email

Commit:
allow to change display names in the user settings again
keep display name and email address in sync with the accounts table

DeepDiver1975 committed:
Bump version

DeepDiver1975 commented:
rebased ...

PVince81 commented:
We're past feature freeze, moving to 9.2.

CC @cmonteroluque

PVince81 commented:
Is there still interest in getting this feature finished ?

PVince81 commented:
Needs to be revived at some point -> backlog for now

cc @pmaier1

DeepDiver1975 commented:
bye bye
